src/array/cpu/labor_pick.h

```diff
@@ -0,0 +1,324 @@
+  const IdxType* rows_data = rows.Ptr<IdxType>();
+  const IdxType* nids = IsNullArray(NIDs) ? nullptr : NIDs.Ptr<IdxType>();
+  const auto num_rows = rows->shape[0];
+  const auto& ctx = mat.indptr->ctx;
+
+  const bool weights = !IsNullArray(prob);
+  // O(1) c computation not possible, so one more iteration is needed.
+  if (importance_sampling >= 0)
+    importance_sampling += weights;
+  // A stands for A in arXiv:2210.13339, that is edge weights
+  auto A_arr = prob;
+  FloatType *A = A_arr.Ptr<FloatType>();
+  constexpr FloatType ONE = 1;
+
+  constexpr auto fidtype = DGLDataTypeTraits<FloatType>::dtype;
+
+  // cs stands for c_s in arXiv:2210.13339
+  FloatArray cs_array = NDArray::Empty({num_rows}, fidtype, ctx);
+  FloatType* cs = cs_array.Ptr<FloatType>();
+  // ds stands for A_{*s} in arXiv:2210.13339
+  FloatArray ds_array = NDArray::Empty({num_rows}, fidtype, ctx);
+  FloatType* ds = ds_array.Ptr<FloatType>();
+
+  IdxType max_degree = 1;
+  IdxType hop_size = 0;
+  for (int64_t i = 0; i < num_rows; ++i) {
+    const IdxType rid = rows_data[i];
+    const auto act_degree = indptr[rid + 1] - indptr[rid];
+    max_degree = std::max(act_degree, max_degree);
+    double d = weights ? std::accumulate(A + indptr[rid], A + indptr[rid + 1], 0.0) : act_degree;
+    // O(1) c computation, samples more than needed for weighted case, mentioned in the sentence
+    // between (10) and (11) in arXiv:2210.13339
+    cs[i] = num_picks / d;
+    ds[i] = d;
+    hop_size += act_degree;
+  }
+
+  phmap::flat_hash_map<IdxType, FloatType> hop_map;
+
+  if (importance_sampling)
+    compute_importance_sampling_probabilities<IdxType, FloatType>(
+        ctx,
+        fidtype,
+        max_degree,
+        num_rows,
+        importance_sampling,
+        weights,
+        rows_data,
+        indptr,
+        A,
+        indices,
+        (IdxType)num_picks,
+        ds,
+        cs,
+        hop_map);
+
+  const uint64_t random_seed = IsNullArray(random_seed_arr) ?
+      RandomEngine::ThreadLocal()->RandInt(1000000000) : random_seed_arr.Ptr<int64_t>()[0];
+
+  // compute number of edges first and store randoms
+  IdxType num_edges = 0;
+  FloatArray rands_arr = NDArray::Empty({hop_size}, fidtype, ctx);
+  auto rands = rands_arr.Ptr<FloatType>();
+  for (int64_t i = 0, off = 0; i < num_rows; i++) {
+    const IdxType rid = rows_data[i];
+    const auto c = cs[i];
+    for (auto j = indptr[rid]; j < indptr[rid + 1]; j++) {
+      const auto v = indices[j];
+      const auto t = nids ? nids[v] : v;  // t in the paper
+      pcg32 ng(random_seed, t);
+      std::uniform_real_distribution<FloatType> uni;
+      // rolled random number r_t is a function of the random_seed and t
+      const auto rnd = uni(ng);
+      // if hop_map is initialized, get ps from there, otherwise get it from the alternative.
+      num_edges +=
+          rnd <= (importance_sampling - weights ? c * hop_map[v] : c * (weights ? A[j] : 1));
+      rands[off++] = rnd;
```

Comment:
Seems that you are precomputing the generated rolls.  Why not just generate them on-the-fly in L269?  This way you don't need to save an array, and the time complexity should be the same.

Reply:
I could generate them on the fly in the next loop also, but calls to the PRNG are expensive compared to array accesses. So this improves the performance slightly.

Reply:
I need to test it again to make sure this is faster though.

Reply:
Technically each element in `rands` array is written and read exactly once, so the number of PRNG calls should be the same.  So generating the numbers on the fly in L269 should save us from allocating `rands_arr` and writing to it.

However, if the current implementation is indeed noticeably faster, you can leave a comment in the code to tell people so that they (including myself) don't try to "optimize" it.  I can't think of a reason why the current implementation could be faster though (other than some caching happening around).

Reply:
Before, the calls to the PRNG were slower when I was using the skip ahead with the vertex id t which had O(log t) complexity. Now though, I decided to pass t as the stream argument, so it should be O(1). Now, not using an array could potentially be faster as I had made this optimization when PRNG calls were logarithmic complexity. Will test and get back here. Will resolve this conversation after changing to whatever is faster and putting a comment if needed.

Reply:
I have gotten rid of the two steps to compute the number of sampled edges and then allocating an array of that size. Doing it in a single step makes the code around 40% faster when importance_sampling=0 and there are no weights. Other scenarios should also be faster now.